# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
path = os.getcwd()
print(path)

C:\Users\j9903\Desktop\SpringBoard\Assignments\Capstone Project 1


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df0 = pd.read_csv(path + '\\data\\step3_output.csv')

In [30]:
df0.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         2  
4              104.0            80.0         2  

[5 rows x 26 columns]

In [4]:
df0.columns

Index(['Unnamed: 0', 'Name', 'state', 'summit_elev', 'vertical_drop', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [5]:
df0 = df0.drop('Unnamed: 0', axis=1)

In [6]:
df0.describe()

summit_elev  vertical_drop       trams   fastEight   fastSixes  \
count    274.000000     274.000000  274.000000  274.000000  274.000000   
mean    4682.463504    1243.058394    0.167883    0.003650    0.193431   
std     3744.095085     941.601288    0.542270    0.060412    0.681380   
min      315.000000     100.000000    0.000000    0.000000    0.000000   
25%     1476.250000     500.000000    0.000000    0.000000    0.000000   
50%     3175.000000    1000.000000    0.000000    0.000000    0.000000   
75%     8009.000000    1800.000000    0.000000    0.000000    0.000000   
max    13150.000000    4425.000000    4.000000    1.000000    6.000000   

        fastQuads        quad      triple      double     surface  ...  \
count  274.000000  274.000000  274.000000  274.000000  274.000000  ...   
mean     1.065693    1.007299    1.551095    1.919708    2.656934  ...   
std      2.187893    1.369454    1.577990    1.894888    2.075279  ...   
min      0.000000    0.000000    0.000000    0.000000    0.000000  ...   
25%      0.000000    0.000000    0.000000    1.000000    1.000000  ...   
50%      0.000000    1.000000    1.000000    1.000000    2.000000  ...   
75%      1.000000    2.000000    2.000000    3.000000    3.000000  ...   
max     15.000000    8.000000    7.000000   14.000000   15.000000  ...   

       SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear   yearsOpen  \
count         274.000000      274.000000        274.000000  274.000000   
mean          687.841615      182.897142        107.277372   59.945255   
std          1136.446732      255.224663         43.627667   14.307479   
min             8.000000        2.000000          0.000000   22.000000   
25%            96.250000       60.000000         91.000000   52.000000   
50%           212.500000      141.500000        110.500000   58.000000   
75%           739.801223      200.000000        133.750000   70.000000   
max          8464.000000     3379.000000        305.000000  104.000000   

       averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
count       274.000000    274.000000    274.000000         274.000000   
mean        187.381872     58.727226     65.259902         120.182482   
std         138.213904     22.711643     21.126427          30.982855   
min          18.000000     17.000000     20.000000          30.000000   
25%          69.250000     45.000000     50.000000         100.000000   
50%         150.000000     57.916957     64.166810         120.000000   
75%         300.000000     68.500000     74.000000         139.750000   
max         669.000000    179.000000    179.000000         305.000000   

       NightSkiing_ac    clusters  
count      274.000000  274.000000  
mean        61.186131    0.565693  
std         99.379789    0.768705  
min          0.000000    0.000000  
25%          0.000000    0.000000  
50%         30.000000    0.000000  
75%         86.000000    1.000000  
max        650.000000    2.000000  

[8 rows x 24 columns]

In [7]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 26 columns):
Name                 274 non-null object
state                274 non-null object
summit_elev          274 non-null int64
vertical_drop        274 non-null int64
trams                274 non-null int64
fastEight            274 non-null float64
fastSixes            274 non-null int64
fastQuads            274 non-null int64
quad                 274 non-null int64
triple               274 non-null int64
double               274 non-null int64
surface              274 non-null int64
total_chairs         274 non-null int64
Runs                 274 non-null float64
TerrainParks         274 non-null float64
LongestRun_mi        274 non-null float64
SkiableTerrain_ac    274 non-null float64
Snow Making_ac       274 non-null float64
daysOpenLastYear     274 non-null float64
yearsOpen            274 non-null float64
averageSnowfall      274 non-null float64
AdultWeekday         274 non-null f

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [8]:
df = pd.concat([df0.drop('state', axis=1), pd.get_dummies(df0["state"])], axis=1)

In [9]:
df.head(10)

Name  summit_elev  vertical_drop  trams  \
0                 Alyeska Resort         3939           2500      1   
1            Eaglecrest Ski Area         2600           1540      0   
2               Hilltop Ski Area         2090            294      0   
3               Arizona Snowbowl        11500           2300      0   
4            Sunrise Park Resort        11100           1800      0   
5  Yosemite Ski & Snowboard Area         7800            600      0   
6                  Bear Mountain         8805           1665      0   
7                    Bear Valley         8500           1900      0   
8         Boreal Mountain Resort         7700            500      0   
9                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          2     2       0       0  ...             0   
1        0.0          0          0     0       0       4  ...             0   
2        0.0          0          0     0       1       0  ...             0   
3        0.0          1          0     2       2       1  ...             0   
4        0.0          0          1     2       3       1  ...             0   
5        0.0          0          0     0       1       3  ...             0   
6        0.0          0          2     1       2       3  ...             0   
7        0.0          1          1     0       2       4  ...             0   
8        0.0          0          1     1       3       1  ...             0   
9        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   
5             0          0     0        0         0           0   
6             0          0     0        0         0           0   
7             0          0     0        0         0           0   
8             0          0     0        0         0           0   
9             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  
5              0          0        0  
6              0          0        0  
7              0          0        0  
8              0          0        0  
9              0          0        0  

[10 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [10]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

In [11]:
X.shape

(274, 58)

In [12]:
X_scaled.shape

(274, 58)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [13]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [14]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [15]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [16]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test,y_pred)

0.7804742850322437

In [17]:
mean_absolute_error(y_test,y_pred)

5.5260722773047855

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [18]:
lm.intercept_

65.80222383454492

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [19]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
total_chairs       1.093721e+13
fastQuads          4.063072e+12
surface            3.853940e+12
double             3.518940e+12
triple             2.930439e+12
quad               2.543173e+12
New York           2.442685e+12
Michigan           2.343347e+12
Colorado           2.291388e+12
California         2.291388e+12
Pennsylvania       2.182343e+12
Wisconsin          2.003836e+12
New Hampshire      1.939623e+12
Minnesota          1.872659e+12
Vermont            1.872659e+12
Utah               1.729183e+12
Montana            1.570041e+12
Oregon             1.570041e+12
Washington         1.570041e+12
New Mexico         1.483039e+12
Idaho              1.389861e+12
Maine              1.389861e+12
Massachusetts      1.389861e+12
fastSixes          1.265371e+12
Wyoming            1.179038e+12
Connecticut        1.179038e+12
North Carolina     1.179038e+12
Nevada             1.056522e+12
Ohio               1.056522e+12
Illinois           1.056522e+12
West Virginia      1.056522e+12
Virginia           1.056522e+12
trams              1.007033e+12
Alaska             9.166678e+11
Arizona            7.498358e+11
South Dakota       7.498358e+11
Indiana            7.498358e+11
Iowa               5.311877e+11
New Jersey         5.311877e+11
Missouri           5.311877e+11
Rhode Island       5.311877e+11
Maryland           5.311877e+11
Tennessee          5.311877e+11
fastEight          1.121897e+11
AdultWeekday       1.901221e+01
summit_elev        1.594003e+00
clusters           1.593018e+00
averageSnowfall    1.222778e+00
Snow Making_ac     1.171631e+00
NightSkiing_ac     1.034058e+00
Runs               6.422119e-01
SkiableTerrain_ac  5.927734e-01
TerrainParks       4.543457e-01
projectedDaysOpen  4.372559e-01
vertical_drop      3.930869e-01
yearsOpen          3.885498e-01
daysOpenLastYear   1.728516e-01
LongestRun_mi      5.812073e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [20]:
df_m2 = df0.copy()

In [21]:
X_m2 = df_m2.drop(['Name','state','AdultWeekend'], axis=1)
y_m2 = df_m2.AdultWeekend
scaler_m2 = preprocessing.StandardScaler().fit(X_m2)
X_scaled_m2 =scaler_m2.transform(X_m2)
y_m2 = y_m2.ravel()
X_train_m2, X_test_m2, y_train_m2, y_test_m2 = train_test_split(X_scaled_m2, y_m2, test_size=0.25, random_state=1)

In [22]:
lm_m2 = linear_model.LinearRegression()
model_m2 = lm_m2.fit(X_train_m2,y_train_m2)
y_pred_m2 = model_m2.predict(X_test_m2)

In [23]:
print("explained_variance_score = " + str(explained_variance_score(y_test_m2,y_pred_m2)))
print("mean_absolute_error = " + str(mean_absolute_error(y_test_m2,y_pred_m2)))
print("Intercept of the model = " + str(lm_m2.intercept_))

explained_variance_score = 0.7703416899459896
mean_absolute_error = 6.352928045607559
Intercept of the model = 66.10695062628587


In [24]:
pd.DataFrame(abs(lm_m2.coef_), X_m2.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday         19.597540
clusters              3.692650
vertical_drop         1.941053
summit_elev           1.465687
fastQuads             1.430344
quad                  1.234816
Runs                  1.211361
triple                1.148021
Snow Making_ac        1.079869
total_chairs          0.822313
averageSnowfall       0.712582
surface               0.704220
fastSixes             0.627321
NightSkiing_ac        0.516928
yearsOpen             0.454343
LongestRun_mi         0.387949
double                0.328769
daysOpenLastYear      0.242769
trams                 0.221411
fastEight             0.198154
projectedDaysOpen     0.135209
TerrainParks          0.057625
SkiableTerrain_ac     0.056303

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [25]:
df_m3 = df0.copy()

In [26]:
X_m3 = df_m3.drop(['Name','state','AdultWeekend','summit_elev'], axis=1)
y_m3 = df_m3.AdultWeekend
scaler_m3 = preprocessing.StandardScaler().fit(X_m3)
X_scaled_m3 =scaler_m3.transform(X_m3)
y_m3 = y_m3.ravel()
X_train_m3, X_test_m3, y_train_m3, y_test_m3 = train_test_split(X_scaled_m3, y_m3, test_size=0.25, random_state=1)

In [27]:
lm_m3 = linear_model.LinearRegression()
model_m3 = lm_m3.fit(X_train_m3,y_train_m3)
y_pred_m3 = model_m3.predict(X_test_m3)

In [28]:
print("explained_variance_score = " + str(explained_variance_score(y_test_m3,y_pred_m3)))
print("mean_absolute_error = " + str(mean_absolute_error(y_test_m3,y_pred_m3)))
print("Intercept of the model = " + str(lm_m3.intercept_))

explained_variance_score = 0.772091293051848
mean_absolute_error = 6.307865270026756
Intercept of the model = 66.0803126135947


In [29]:
pd.DataFrame(abs(lm_m3.coef_), X_m3.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday         19.607377
clusters              2.472132
vertical_drop         2.123730
fastQuads             1.361400
Snow Making_ac        1.235866
quad                  1.227765
triple                1.184404
Runs                  0.963433
total_chairs          0.811508
surface               0.701918
fastSixes             0.683102
averageSnowfall       0.595751
NightSkiing_ac        0.478360
yearsOpen             0.425177
double                0.313330
LongestRun_mi         0.305583
trams                 0.274142
fastEight             0.264261
daysOpenLastYear      0.208252
SkiableTerrain_ac     0.175348
projectedDaysOpen     0.148848
TerrainParks          0.044575

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.7804742850322437 | 5.5260722773047855 |'base_elev'|
| Model 2. | 0.7703416899459896 | 6.352928045607559 |'state','base_elev'|
| Model 3. | 0.772091293051848 | 6.307865270026756 |'state','summit_elev','base_elev'|

Model Selection: Model 1

The success criteria of the model performance is 

1. Smaller MAE (ideally <= 40; small MAE suggests the model is great at prediction, while a large MAE suggests that your model may have trouble in certain areas. A MAE of 0 means that your model is a perfect predictor of the outputs but this will almost never happen)
2. Higher EV (for EV, the best possible score is 1.0 and lower values are worse)

Therefore model 2 and 3 are not selected because their MAE values are higher and EV values are lower than model 1.